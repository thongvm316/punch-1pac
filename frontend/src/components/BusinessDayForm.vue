<template>
  <div>
    <div class="form-group" :class="{ 'has-error': $v.params.weekday.$error || errors.weekday }">
      <label class="form-label">{{ $t('company.businessDays.labels.weekday') }}</label>
      <select class="form-select" v-model="$v.params.weekday.$model">
        <option value="">{{ $t('company.businessDays.placeholder.chooseWeekday') }}</option>
        <option :value="weekday" v-for="(weekday, key) in meta.weekdays" :key="key">{{ $t(`meta.weekdays.${weekday}`) }}</option>
      </select>
      <p class="form-input-hint" v-if="$v.params.weekday.$error">
        {{ $t('validation.required', { name: $t('company.businessDays.labels.weekday') }) }}
      </p>
      <p class="form-input-hint" v-if="errors.weekday">{{ $t('company.businessDays.labels.weekday') }} {{ errors.weekday[0] }}</p>
    </div>
    <div class="form-group" :class="{ 'has-error': $v.params.morning_started_at.$error || errors.morning_started_at }">
      <label class="form-label">{{ $t('company.businessDays.labels.morningStartAt') }}</label>
      <input class="form-input" type="time" step="300" v-model="$v.params.morning_started_at.$model">
      <p class="form-input-hint" v-if="$v.params.morning_started_at.$error">
        {{ $t('validation.required', { name: $t('company.businessDays.labels.morningStartAt') }) }}
      </p>
      <p class="form-input-hint" v-if="errors.morning_started_at">{{ $t('company.businessDays.labels.morningStartAt') }} {{ errors.morning_started_at[0] }}</p>
    </div>
    <div class="form-group" :class="{ 'has-error': $v.params.morning_ended_at.$error || errors.morning_ended_at }">
      <label class="form-label">{{ $t('company.businessDays.labels.morningEndAt') }}</label>
      <input class="form-input" type="time" step="300" v-model="$v.params.morning_ended_at.$model">
      <p class="form-input-hint" v-if="$v.params.morning_ended_at.$error">
        {{ $t('validation.required', { name: $t('company.businessDays.labels.morningEndAt') }) }}
      </p>
      <p class="form-input-hint" v-if="errors.morning_ended_at">{{ $t('company.businessDays.labels.morningEndAt') }} {{ errors.morning_ended_at[0] }}</p>
    </div>
    <div class="form-group" :class="{ 'has-error': $v.params.afternoon_started_at.$error || errors.afternoon_started_at }">
      <label class="form-label">{{ $t('company.businessDays.labels.afternoonStartAt') }}</label>
      <input class="form-input" type="time" step="300" v-model="$v.params.afternoon_started_at.$model">
      <p class="form-input-hint" v-if="$v.params.afternoon_started_at.$error">
        {{ $t('validation.required', { name: $t('company.businessDays.labels.afternoonStartAt') }) }}
      </p>
      <p class="form-input-hint" v-if="errors.afternoon_started_at">{{ $t('company.businessDays.labels.afternoonStartAt') }} {{ errors.afternoon_started_at[0] }}</p>
    </div>
    <div class="form-group" :class="{ 'has-error': $v.params.afternoon_ended_at.$error || errors.afternoon_ended_at }">
      <label class="form-label">{{ $t('company.businessDays.labels.afternoonEndAt') }}</label>
      <input class="form-input" type="time" step="300" v-model="$v.params.afternoon_ended_at.$model">
      <p class="form-input-hint" v-if="$v.params.afternoon_ended_at.$error">
        {{ $t('validation.required', { name: $t('company.businessDays.labels.afternoonEndAt') }) }}
      </p>
      <p class="form-input-hint" v-if="errors.afternoon_ended_at">{{ $t('company.businessDays.labels.afternoonEndAt') }} {{ errors.afternoon_ended_at[0] }}</p>
    </div>
    <div class="form-group">
      <button
        ref="createBusinessDayButton"
        type="button"
        class="btn btn-success btn-submit"
        v-if="!targetBusinessDay"
        @click="localAddBusinessDay"
        :disabled="isDisabled">
        {{ $t('company.businessDays.btn.submit') }}
      </button>
      <button
        ref="editBusinessDayButton"
        type="button"
        class="btn btn-success btn-submit"
        v-if="targetBusinessDay"
        @click="localEditBusinessDay"
        :disabled="isDisabled">
        {{ $t('company.businessDays.btn.save') }}
      </button>
    </div>
  </div>
</template>

<script>
import { CLEAR_BUSINESS_DAY_ERRORS } from '../store/mutation-types'
import { mapState, mapActions, mapMutations } from 'vuex'
import { isEmpty, isEqual } from 'underscore'
import handleSuccess from '../mixins/handle-success'
import businessDayValidate from '../validations/business-day-validate'

export default {
  name: 'business-day-form',

  mixins: [handleSuccess, businessDayValidate],

  props: {
    targetBusinessDay: Object
  },

  data() {
    return {
      params: {
        weekday: '',
        morning_started_at: '',
        morning_ended_at: '',
        afternoon_started_at: '',
        afternoon_ended_at: ''
      },
      data: {
        emitType: 'afterModify',
        message: ''
      }
    }
  },

  methods: {
    ...mapActions('companyBusinessDays', ['addBusinessDay', 'updateBusinessDay']),

    ...mapMutations('companyBusinessDays', [CLEAR_BUSINESS_DAY_ERRORS]),

    localAddBusinessDay() {
      this.addBusinessDay(this.params).then(response => {
        this.data.message = this.$t('messages.businessDay.createSuccess')
        this.handleSuccess(this.data)
      })
    },

    localEditBusinessDay() {
      this.updateBusinessDay({ businessDayId: this.targetBusinessDay.id, updateParams: this.params }).then(response => {
        this.data.message = this.$t('messages.businessDay.updateSuccess')
        this.handleSuccess(this.data)
      })
    }
  },

  computed: {
    ...mapState('initialStates', ['meta']),

    ...mapState('companyBusinessDays', ['errors']),

    isDisabled() {
      if (this.$v.params.$anyError) return true

      let flag = false
      if (this.targetBusinessDay) {
        flag = isEqual(this.params, this.targetBusinessDay)
      } else {
        const emptyParams = {
          weekday: '',
          morning_started_at: '',
          morning_ended_at: '',
          afternoon_started_at: '',
          afternoon_ended_at: ''
        }

        flag = isEqual(this.params, emptyParams)
      }
      return flag
    }
  },

  created() {
    if (!isEmpty(this.errors)) this[CLEAR_BUSINESS_DAY_ERRORS]()
    if (this.targetBusinessDay) this.params = { ...this.targetBusinessDay }
  }
}
</script>
