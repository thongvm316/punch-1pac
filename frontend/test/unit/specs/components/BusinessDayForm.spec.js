import { shallowMount } from '@vue/test-utils'

import wrapperOps from '../../supports/wrapper'
import setComputed from '../../supports/set-computed'

import BusinessDayForm from '@/components/BusinessDayForm'

const localAddBusinessDay = jest.fn()
const localEditBusinessDay = jest.fn()

describe('BusinessDayForm.vue', () => {
  let wrapper

  afterEach(() => {
    wrapper.vm.$destroy()
  })

  describe('when BusinessDayForm was mounted', () => {
    beforeEach(() => {
      Object.assign(wrapperOps, { methods: { localAddBusinessDay } })
      wrapper = shallowMount(BusinessDayForm, wrapperOps)
    })

    it('should display AllowedIpForm Component', () => {
      expect(wrapper.isVueInstance()).toBe(true)
      expect(wrapper.exists()).toBe(true)
    })

    it('validation prop data', () => {
      const { targetBusinessDay } = wrapper.vm.$options.props
      expect(targetBusinessDay.required).toBeFalsy()
      expect(targetBusinessDay.type).toBe(Object)
    })

    it('should display create business day button', () => {
      expect(wrapper.find({ ref: 'createBusinessDayButton' }).exists()).toBe(true)
      expect(wrapper.find({ ref: 'editBusinessDayButton' }).exists()).toBe(false)
    })

    it('localAddBusinessDay method should have been call', async () => {
      wrapper.find({ ref: 'createBusinessDayButton' }).trigger('click')
      await wrapper.vm.$nextTick()

      expect(localAddBusinessDay).toHaveBeenCalled()
    })

    it('should show error text', async () => {
      setComputed(wrapper, {
        errors: {
          weekday: ['can\'t be empty'],
          morning_started_at: ['can\'t be empty'],
          morning_ended_at: ['can\'t be empty'],
          afternoon_started_at: ['can\'t be empty'],
          afternoon_ended_at: ['can\'t be empty']
        }
      })
      await wrapper.vm.$nextTick()

      const formGroups = wrapper.findAll('.form-group')
      const formInputErrors = wrapper.findAll('.form-group > .form-input-hint')

      for (let i = 0; i < formInputErrors.length; i++) {
        expect(formGroups.at(i).classes()).toContain('has-error')
        expect(formGroups.at(i).find('.form-input-hint').exists()).toBe(true)
        expect(formGroups.at(i).find('.form-input-hint').text()).toContain('can\'t be empty')
      }
    })
  })

  describe('when BusinessDayForm was mounted with props', () => {
    beforeEach(() => {
      Object.assign(wrapperOps, {
        propsData: {
          targetBusinessDay: {
            weekday: 'monday',
            morning_started_at: '8:00',
            morning_ended_at: '8:00',
            afternoon_started_at: '8:00',
            afternoon_ended_at: '8:00'
          }
        },
        methods: {
          localEditBusinessDay
        }
      })
      wrapper = shallowMount(BusinessDayForm, wrapperOps)
    })

    it('should display update business day button', () => {
      expect(wrapper.find({ ref: 'createBusinessDayButton' }).exists()).toBe(false)
      expect(wrapper.find({ ref: 'editBusinessDayButton' }).exists()).toBe(true)
    })

    it('should update data from props targetBusinessDay', () => {
      const dataParams = wrapper.vm.$data.params
      const propsParams = wrapper.vm.$props.targetBusinessDay

      expect(propsParams.weekday).toBe(dataParams.weekday)
      expect(propsParams.morning_started_at).toBe(dataParams.morning_started_at)
      expect(propsParams.morning_ended_at).toBe(dataParams.morning_ended_at)
      expect(propsParams.afternoon_started_at).toBe(dataParams.afternoon_started_at)
      expect(propsParams.afternoon_ended_at).toBe(dataParams.afternoon_ended_at)
    })

    it('localEditBusinessDay method should have been call', async () => {
      wrapper.find({ ref: 'editBusinessDayButton' }).trigger('click')
      await wrapper.vm.$nextTick()

      expect(localEditBusinessDay).toHaveBeenCalled()
    })
  })
})
