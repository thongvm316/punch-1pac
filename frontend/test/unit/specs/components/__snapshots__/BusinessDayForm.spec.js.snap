// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BusinessDayForm.vue when have no props when render correctly should display AllowedIpForm Component 1`] = `
<div>
  <div class="form-group"><label class="form-label">Weekday</label> <select class="form-select">
      <option value="">
        Choose a weekday
      </option>
      <option value="monday">
        Monday
      </option>
      <option value="tuesday">
        Tuesday
      </option>
      <option value="wednesday">
        Wednesday
      </option>
      <option value="thursday">
        Thursday
      </option>
      <option value="friday">
        Friday
      </option>
      <option value="saturday">
        Saturday
      </option>
      <option value="sunday">
        Sunday
      </option>
    </select>
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Morning start at</label>
    <flatpickr-stub config="[object Object]" value="08:00" class="form-input time-picker"></flatpickr-stub>
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Morning end at</label>
    <flatpickr-stub config="[object Object]" value="12:00" class="form-input time-picker"></flatpickr-stub>
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Afternoon start at</label>
    <flatpickr-stub config="[object Object]" value="13:30" class="form-input time-picker"></flatpickr-stub>
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Afternoon end at</label>
    <flatpickr-stub config="[object Object]" value="17:30" class="form-input time-picker"></flatpickr-stub>
    <!---->
    <!---->
  </div>
  <div class="form-group"><button type="button" class="btn btn-success btn-submit">
      Submit
    </button>
    <!---->
  </div>
</div>
`;

exports[`BusinessDayForm.vue when have no props when render correctly should show error text 1`] = `
<div>
  <div class="form-group has-error"><label class="form-label">Weekday</label> <select class="form-select">
      <option value="">
        Choose a weekday
      </option>
      <option value="monday">
        Monday
      </option>
      <option value="tuesday">
        Tuesday
      </option>
      <option value="wednesday">
        Wednesday
      </option>
      <option value="thursday">
        Thursday
      </option>
      <option value="friday">
        Friday
      </option>
      <option value="saturday">
        Saturday
      </option>
      <option value="sunday">
        Sunday
      </option>
    </select>
    <!---->
    <p class="form-input-hint">
      Weekday không thể để trắng
    </p>
  </div>
  <div class="form-group"><label class="form-label">Morning start at</label>
    <flatpickr-stub config="[object Object]" value="08:00" class="form-input time-picker"></flatpickr-stub>
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Morning end at</label>
    <flatpickr-stub config="[object Object]" value="12:00" class="form-input time-picker"></flatpickr-stub>
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Afternoon start at</label>
    <flatpickr-stub config="[object Object]" value="13:30" class="form-input time-picker"></flatpickr-stub>
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Afternoon end at</label>
    <flatpickr-stub config="[object Object]" value="17:30" class="form-input time-picker"></flatpickr-stub>
    <!---->
    <!---->
  </div>
  <div class="form-group"><button type="button" class="btn btn-success btn-submit">
      Submit
    </button>
    <!---->
  </div>
</div>
`;

exports[`BusinessDayForm.vue when props targetBusinessDay when was mouted should render correctly 1`] = `
<div>
  <div class="form-group"><label class="form-label">Weekday</label> <select class="form-select">
      <option value="">
        Choose a weekday
      </option>
      <option value="monday">
        Monday
      </option>
      <option value="tuesday">
        Tuesday
      </option>
      <option value="wednesday">
        Wednesday
      </option>
      <option value="thursday">
        Thursday
      </option>
      <option value="friday">
        Friday
      </option>
      <option value="saturday">
        Saturday
      </option>
      <option value="sunday">
        Sunday
      </option>
    </select>
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Morning start at</label>
    <flatpickr-stub config="[object Object]" value="8:00" class="form-input time-picker"></flatpickr-stub>
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Morning end at</label>
    <flatpickr-stub config="[object Object]" value="12:00" class="form-input time-picker"></flatpickr-stub>
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Afternoon start at</label>
    <flatpickr-stub config="[object Object]" value="13:00" class="form-input time-picker"></flatpickr-stub>
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Afternoon end at</label>
    <flatpickr-stub config="[object Object]" value="17:00" class="form-input time-picker"></flatpickr-stub>
    <!---->
    <!---->
  </div>
  <div class="form-group">
    <!----> <button type="button" class="btn btn-success btn-submit">
      Save
    </button></div>
</div>
`;
